### ISSUE TITLE
[BUG] Webkit does not properly navigate after onSubmit (NextJS App Router)

### ISSUE THREAD

**Issue by qwerqy** [microsoft/playwright#26091]
### System info
- Playwright Version: v1.36.2
- Operating System: macOS 13.2
- Browser: WebKit
- Other info: NextJS 13.4, react-hook-form

### Source code

Here is my search form:

```tsx
'use client'

import {
	useParams,
	usePathname,
	useRouter,
	useSearchParams,
} from 'next/navigation'
import { useEffect, useRef, useTransition } from 'react'
import { useForm } from 'react-hook-form'

export default function SearchForm() {
	const router = useRouter()
	const pathname = usePathname()
	const params = useParams()
	const searchParams = useSearchParams()
	const initialQuery = params?.query
	const { register, handleSubmit, reset, setValue } = useForm<{
		query: string
	}>({})

	const [isPending, setTransition] = useTransition()

	const onSearch = (values: { query: string }) => {
		const { query } = values

		setTransition(() => {
			router.push(
				`/search/${encodeURIComponent(query)}` +
					(searchParams ? '?' + searchParams?.toString() : ''),
			)
		})
	}

	useEffect(() => {
		if (initialQuery) {
			setValue('query', decodeURIComponent(initialQuery as string))
		}
	}, [initialQuery, setValue])

	useEffect(() => {
		if (!pathname?.includes('search')) {
			reset()
		}
	}, [pathname, reset])

	return (
		<form
			onSubmit={handleSubmit(onSearch)}
			action={undefined}
			className="w-full max-w-5xl mx-auto"
		>
			<div className="join w-full">
				<input
					type="text"
					className="input input-bordered w-full join-item"
					placeholder="Search for a ☕️ name or location"
					{...register('query')}
				/>
				<button type="submit" className="btn btn-secondary join-item">
					{isPending ? <span className="loading loading-spinner"></span> : '🔎'}
				</button>
			</div>
		</form>
	)
}
```

Here is my E2E test:

```tsx
import { test, expect } from '@playwright/test'

test.only('Find a cafe', async ({ page }) => {
	await page.goto('/')
	const searchInput = page.locator('input[name="query"]')
	expect(searchInput).toBeVisible()

	await searchInput.fill('cafe')
	await expect(searchInput.inputValue()).resolves.toEqual('cafe')
	await searchInput.press('Enter')

	await page.waitForURL('/search/cafe')
	expect(page.url()).toContain('/search/cafe')

	await expect(
		page.getByRole('heading', {
			name: 'Search through 4000+ awesome cafes around Klang Valley!',
		}),
	).toBeVisible()

	await expect(page.getByTestId('query-search-pill')).toBeVisible()

	await expect(page.getByText(/Showing [0-9]+ results/i)).toBeVisible()

	// Get data-testid cafe-item and count the elements to equal 15
	await expect(page.locator('[data-testid="cafe-item"]').count()).resolves.toBe(
		15,
	)
})
```

The error I get from running `npx playwright test`

```
Test timeout of 30000ms exceeded.

    Error: page.waitForURL: Navigation failed because page was closed!
    =========================== logs ===========================
    waiting for navigation to "/search/cafe" until "load"
      navigated to "http://localhost:3000/?query=cafe"
      navigated to "http://localhost:3000/?query=cafe"
    ============================================================

      10 |      await searchInput.press('Enter')
      11 |
    > 12 |      await page.waitForURL('/search/cafe')
         |                 ^
      13 |      expect(page.url()).toContain('/search/cafe')
      14 |
      15 |      await expect(
```

**NOTE**: Running on --ui works, this error only happens when running on headless
**NOTE2**: I screenshot the page after the enter button is pressed and on the screenshot, the input did not capture which is weird because I have an assertion after I typed in `cafe` in the input. 

**Steps**
- Run `npx playwright test`

**Expected**

Test should pass because it passes on Chromium and Firefox

**Actual**

It seems that in webkit, after clicking submit, it calls the action of Form (which is why the navigate URL was `/search?query`.
I have tried searching for solutions online for 2 days before I decided to post here.

**Comment by qwerqy** [comment]
Server logs from Firefox and Chromium tests
```
-  ┌ GET / 200 in 1060ms
   │
   ├──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafes?select=id%2Cnam.. 206 in 867ms (cache: MISS)
   │
   └── 1 level ── GET https://pgrcpqjkquvvtsnu../rest/v1/cafe_count?select=* 200 in 73ms (cache: MISS)

-  ┌ GET /search/cafe?_rsc=587a9e7 200 in 275ms
   │
   ├──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafe_count?select=* 200 in 101ms (cache: MISS)
   │
   └──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafes?select=id%2Cnam.. 206 in 210ms (cache: MISS)


-  ┌ GET /search/cafe?_rsc=587a9e7 200 in 256ms
   │
   ├──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafe_count?select=* 200 in 56ms (cache: MISS)
   │
   └──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafes?select=id%2Cnam.. 206 in 222ms (cache: MISS)
```

Server logs for Webkit test
```
-  ┌ GET / 200 in 1079ms
   │
   ├──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafes?select=id%2Cnam.. 206 in 931ms (cache: MISS)
   │
   └── 1 level ── GET https://pgrcpqjkquvvtsnu../rest/v1/cafe_count?select=* 200 in 55ms (cache: MISS)

false
/

-  ┌ GET /?query=cafe 200 in 1021ms
   │
   ├──── GET https://pgrcpqjkquvvtsnu../rest/v1/cafes?select=id%2Cnam.. 206 in 874ms (cache: MISS)
   │
   └── 1 level ── GET https://pgrcpqjkquvvtsnu../rest/v1/cafe_count?select=* 200 in 61ms (cache: MISS)

```

**Comment by dgozman** [comment]
@qwerqy Thank you for filing! We appreciate the repro and logs, but unfortunately we cannot run it verbatim since there's so much missing. Would it be possible for you to create a sample GitHub repository that includes all the necessary files so that we can `npm install && npm start` it? Otherwise we won't be able to debug and fix the issue.

> It seems that in webkit, after clicking submit, it calls the action of Form (which is why the navigate URL was /search?query.

This could be the case. Note that you can open your site in WebKit with `npx playwright wk <page-url>`, then open Web Inspector and try to debug the issue. Since you are familiar with the codebase, that would be much more productive than us trying to figure out the issue.

On a side note, have you tried your site in Safari? Does the issue persist there?

**Comment by qwerqy** [comment]
Hi @dgozman thanks for getting back. I will try to debug it further and hopefully create a working repro soon. I did some research 2 nights ago and i think it could be due to hydration — page loads in succesfully , but javascript loads in a few ms late. 

In detail from what i could deduce, 
- page renders html fast (ssr).
- all the elements are visible without any interactivity (this could explain the form being in its “pure” form).
- playwright types in input and submits form with no js
- Form’s action is called instead of onSubmit (cause it hasnt loaded in)

I have yet done anymore testing because I am currently not working on it until this weekend. And yes, I have tested the site on Safari both on MacOS and iOS and both are working perfect. Even on `npx playwright test --ui` on Webkit, the tests run fine. It is only happening on `headless` mode.
